pkg/controllers/routereflector/topologies/multi.go

```diff
@@ -0,0 +1,318 @@
+		return len(statuses[i].Nodes) < len(statuses[j].Nodes)
+	})
+
+	return
+}
+
+// GenerateBGPPeers generates one mesh for the RRs and a fixed amount of client-RR configs per node
+func (t *MultiTopology) GenerateBGPPeers(affectedNodes map[*corev1.Node]bool, existingPeers []*apiv3.BGPPeer) (toRefresh []*apiv3.BGPPeer, toDelete []*apiv3.BGPPeer) {
+	routeReflectors := t.collectRouteReflectors(affectedNodes)
+
+	// Sorting RRs and Nodes for deterministic RR for Node selection
+	sort.Slice(routeReflectors, func(i, j int) bool {
+		return routeReflectors[i].GetCreationTimestamp().UnixNano() < routeReflectors[j].GetCreationTimestamp().UnixNano()
+	})
+
+	nodeList := []*corev1.Node{}
+	for n := range affectedNodes {
+		nodeList = append(nodeList, n)
+	}
+	// Sorting node list for deterministic node order
+	sort.Slice(nodeList, func(i, j int) bool {
+		return nodeList[i].GetCreationTimestamp().UnixNano() < nodeList[j].GetCreationTimestamp().UnixNano()
+	})
+
+	// Find or generate RR mesh config
+	rrConfig := findBGPPeer(existingPeers, DefaultRouteReflectorMeshName)
+	if rrConfig == nil {
+		log.Debugf("Creating new RR full-mesh BGPPeers: %s", DefaultRouteReflectorMeshName)
+		rrConfig = generateBGPPeerStub(DefaultRouteReflectorMeshName)
+	}
+
+	toKeep := map[string]bool{}
+	selector := fmt.Sprintf("has(%s)", t.NodeLabelKey)
+	// Update if different than expected
+	if rrConfig.Spec.NodeSelector != selector || rrConfig.Spec.PeerSelector != selector {
+		rrConfig.Spec = apiv3.BGPPeerSpec{
+			NodeSelector: selector,
+			PeerSelector: selector,
+		}
+
+		toRefresh = append(toRefresh, rrConfig)
+	} else {
+		toKeep[rrConfig.GetName()] = true
+	}
+
+	// Used for round robin RR selection
+	rrIndex := -1
+	rrIndexPerZone := map[string]int{}
+	zones := []string{}
+	rrPerZone := map[string][]*corev1.Node{}
+
+	// Create per zone RR lists for MZR selection
+	if t.Config.ZoneLabel != "" {
+		for i, rr := range routeReflectors {
+			rrZone := rr.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("RR:%s's zone: %s", rr.GetName(), rrZone)
+			rrPerZone[rrZone] = append(rrPerZone[rrZone], routeReflectors[i])
+		}
+
+		for zone := range rrPerZone {
+			zones = append(zones, zone)
+		}
+		sort.Strings(zones)
+	}
+
+	// Let's find RRs for each nodes
+	for _, n := range nodeList {
+		if t.IsRouteReflector(string(n.GetUID()), n.GetLabels()) {
+			continue
+		}
+
+		routeReflectorsForNode := []*corev1.Node{}
+
+		// Run MZR selection first for nodes if zone label is configured
+		if t.Config.ZoneLabel != "" {
+			nodeZone := n.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("Node's zone: %s", nodeZone)
+
+			// Select the 1st RR from the same zone if there're any
+			if len(rrPerZone[nodeZone]) > 0 {
+				rr := selectRRfromZone(rrIndexPerZone, rrPerZone, nodeZone)
+				log.Debugf("Adding %s as 1st RR for Node:%s", rr.GetName(), n.GetName())
+				routeReflectorsForNode = append(routeReflectorsForNode, rr)
+			}
+
+			// Select the 2nd RR from a different zone
```

Comment:
Hm, this isn't what I originally thought this topology would look like. I was imagining something like this:

- Each "zone" gets its own RR cluster, size of that cluster is determined by number of nodes in the zone.
- Each zone's RRs all peer with each other in a mesh. 
- Nodes in the zone establish peering with all of the RRs in that zone's cluster.

The final step is then peering those disjoint clusters together - than can be done in one of a few ways:
- For on-prem, peering with the ToR in the same AS (potentially a manual step)
- For cloud, peer the RR clusters with each other to share routes. 

Although ideally for cloud where there is no ToR, the user wouldn't need BGP at all and could use something like VXLAN instead of a scaled RR topology.

Either way, I'm not sure about peering nodes to RRs that aren't in the same zone as them - doesn't that just produce a cross-zone dependency, where instead we want to prevent a zone failure from impacting other zones?

Maybe I am misunderstanding though! Could you describe in a bit more detail what the desired BGP topology is for the multi-zone case?